upgrade.md

```diff
@@ -29,65 +25,56 @@ Provided you have suitable udev rules in play, this will already do everything y
 Download or compile a Black Magic Debug (BMD) firmware.
 As with the [bmputil](#bmputil) section above, the same considerations exist with this method for firmware selection.
 
+You will need to install [`dfu-util`](http://dfu-util.sourceforge.net/) package. On macOS you can use homebrew, macports or fink depending on your preferred package manager. You will need version 0.8.0 or greater to support the dfuse commands.
+
 Plug in the Black Magic Probe into your computer and run the following command:
 
 ```bash
 sudo dfu-util -d 1d50:6018,:6017 -s 0x08002000:leave -D blackmagic-native.bin
 ```
 
-You will need to install [`dfu-util`](http://dfu-util.sourceforge.net/) package. On MacOS you can use homebrew, macports or fink depending on your preferred package manager. You will need version 0.8.0 or greater to support the dfuse commands.
+To upgrade non-native hardware see the readme of the different [platforms on GitHub](https://github.com/blackmagic-debug/blackmagic/tree/main/src/platforms).
 
 ```{note}
 If `dfu-util` fails to switch your BMP into bootloader mode, or you feel like you might have **bricked** your BMP, you can also plug in your BMP while holding down the button. This will force the BMP to stay in the bootloader on power up.
 ```
 
-## Windows
+### Using stm32-mem.py
 
-### bmputil
-
-Download or compile a Black Magic Probe (BMP) firmware.
-
-Clone and build, or directly install the bmputil tool:
+An alternative solution is using `stm32_mem.py` script included in the Black Magic Debug source repository. It is not the recommended way of updating the Black Magic Probe as `dfu-util` is easier to install without the need to clone the Black Magic Debug repository and all the associated source code.
 
 ```bash
-cargo install --git https://github.com/blackmagic-debug/bmputil
+cd blackmagic && scripts/stm32_mem.py blackmagic.bin
 ```
 
-Plug in your Black Magic Probe and run the following command:
+### stlink-tool
+
+An ST-Link can be used to upate the firmware using [stlink-tool](<https://github.com/jeanthom/stlink-tool>).
 
 ```bash
-bmputil flash blackmagic-native.elf
+stlink-tools blackmagic.bin
 ```
 
-bmputil will automatically take care of your driver needs as appropriate.
+## Windows (specific)
+
+### bmputil
+For Windows, the bmputil instructions don't change.
```

Comment:
Missing new line between the header and this text (Markdown lint warning).

The "Download or compile" lines (which need `s/a/the/` applying) were added because we kept getting Windows users who did not realise they had to do that first. They were added in response to that, and it stemmed the flow

Reply:
I have updated the sections, such that the h2 sections (##) are used for the different programs, and the h3 sections (###) for the different operating systems, to keep the number of duplicated lines to a minimum.

Reply:
Hmm.. We'd originally picked OS-then-tools because people generally start by looking for instructions based on the OS they run - the duplication was deemed acceptable in light of it being the more natural ordering.

tool-then-OS ordering feels very.. clunky.. to traverse especially as it means people are unlikely to ever look further down than the first section and realise there are additional tool choices applicable to their OS. We unfortunately have to write these instructions for the lowest common denominator user, and that tends to be the Windows user who sees instructions for Windows and doesn't look any further than that first instance

Reply:
Ok, then I'll update the page according to the OS-then-tool method, and replicate some of the contents.

Reply:
This is implemented.